src/mdns.rs

```diff
@@ -0,0 +1,444 @@
+        service_type: impl AsRef<str>,
+        proto: impl AsRef<str>,
+        key: impl AsRef<str>,
+        value: impl AsRef<str>,
+    ) -> Result<(), EspError> {
+        let service_type = CString::new(service_type.as_ref()).unwrap();
+        let proto = CString::new(proto.as_ref()).unwrap();
+        let key = CString::new(key.as_ref()).unwrap();
+        let value = CString::new(value.as_ref()).unwrap();
+
+        esp!(unsafe {
+            mdns_service_txt_item_set(
+                service_type.as_ptr(),
+                proto.as_ptr(),
+                key.as_ptr(),
+                value.as_ptr(),
+            )
+        })
+    }
+
+    pub fn service_txt_item_remove(
+        &mut self,
+        service_type: impl AsRef<str>,
+        proto: impl AsRef<str>,
+        key: impl AsRef<str>,
+    ) -> Result<(), EspError> {
+        let service_type = CString::new(service_type.as_ref()).unwrap();
+        let proto = CString::new(proto.as_ref()).unwrap();
+        let key = CString::new(key.as_ref()).unwrap();
+
+        esp!(unsafe {
+            mdns_service_txt_item_remove(service_type.as_ptr(), proto.as_ptr(), key.as_ptr())
+        })
+    }
+
+    pub fn service_txt_set(
+        &mut self,
+        service_type: impl AsRef<str>,
+        proto: impl AsRef<str>,
+        txt: &[(&str, &str)],
+    ) -> Result<(), EspError> {
+        let service_type = CString::new(service_type.as_ref()).unwrap();
+        let proto = CString::new(proto.as_ref()).unwrap();
+        let mut txtcstr = Vec::with_capacity(txt.len());
+        let mut txtptr = Vec::with_capacity(txt.len());
+        for e in txt.iter() {
+            let key = CString::new(e.0.as_bytes()).unwrap();
+            let value = CString::new(e.1.as_bytes()).unwrap();
+            txtptr.push(mdns_txt_item_t {
+                key: key.as_ptr(),
+                value: value.as_ptr(),
+            });
+            txtcstr.push((key, value));
+        }
+
+        esp!(unsafe {
+            mdns_service_txt_set(
+                service_type.as_ptr(),
+                proto.as_ptr(),
+                txtptr.as_mut_ptr(),
+                txtptr.len() as _,
+            )
+        })
+    }
+
+    pub fn service_remove(
+        &mut self,
+        service_type: impl AsRef<str>,
+        proto: impl AsRef<str>,
+    ) -> Result<(), EspError> {
+        let service_type = CString::new(service_type.as_ref()).unwrap();
+        let proto = CString::new(proto.as_ref()).unwrap();
+
+        esp!(unsafe { mdns_service_remove(service_type.as_ptr(), proto.as_ptr()) })
+    }
+
+    pub fn service_remove_all(&mut self) -> Result<(), EspError> {
+        esp!(unsafe { mdns_service_remove_all() })
+    }
+
+    pub fn query(
+        &mut self,
+        name: Option<&str>,
+        service_type: Option<&str>,
+        proto: Option<&str>,
+        mdns_type: Type,
+        timeout: Duration,
+        max_results: usize,
+    ) -> Result<Vec<QueryResult>, EspError> {
```

Comment:
Given that this method takes `max_result: usize`, let's instead of returning a `Vec<QueryResult>` take a `&mut [QueryResult]` where the method should fill it in. And the method should return `Result<usize, EspError>`, where the `usize` is the number of elements in the slice which were filled in.

Reply:
Can we use heapless::Vec?

Reply:
Sure, though i kind of don't see the benefit? If you go for heapless Vec, `max_result` should become a const generic which defines the size of the heapless vec.

Reply:
that's true.

Reply:
We fixed it.